maitake-sync/src/wait_queue.rs

```diff
@@ -684,6 +684,56 @@ impl WaitQueue {
         }
     }
 
+    /// Asynchronously poll the [`WaitQueue`] until a condition occurs
+    ///
+    /// This can be used to implement a "wait loop", turning a "try" function
+    /// (e.g. "try_recv" or "try_send") into an asynchronous function (e.g.
+    /// "recv" or "send").
+    ///
+    /// Consider using [`Self::wait_for_value()`] if your function does return a value.
+    ///
+    /// * Returns `Ok(())` if the closure returns `true`.
+    /// * Returns `Err(Closed)` if the [`WaitQueue`] is closed.
+    pub async fn wait_for<F: FnMut() -> bool>(&self, mut f: F) -> WaitResult<()> {
+        loop {
+            let wait = self.wait();
+            let mut pwait = core::pin::pin!(wait);
+            match pwait.as_mut().subscribe() {
+                Poll::Ready(wr) => wr?,
+                Poll::Pending => {}
+            }
+            if f() {
+                return Ok(());
+            }
+            pwait.await?;
```

Comment:
nit, take it or leave it: i would personally just shadow the name `wait` instead of using `pwait`, 
```suggestion
            let mut wait = core::pin::pin!(wait);
            match wait.as_mut().subscribe() {
                Poll::Ready(wr) => wr?,
                Poll::Pending => {}
            }
            if f() {
                return Ok(());
            }
            wait.await?;
```
but, not a big deal either way.